apache/trafficcontrol
# Add enhanced X509 Certificate/Private RSA Key validation to Traffic Ops (AddSSLKeys Endpoint)

## Which issue is fixed by this PR? If not related to an existing issue, what does this PR do?

Partially addresses #3267 

All imported x509 certificates / private key pairs will be validated with the following criteria:

1) All PEM encoded strings contained in the CRT and KEY fields must be decodable to DER binary format via pem.Decode()
2) A single PEM encoded private RSA key must be decodable from KEY payload.  KEY format must be PKCS #1 or PKCS #8
3) The import server X509 certificate in the CRT payload is checked for the appropriate Public Key Algorithm (x509.RSA is only supported currently)
4) Private RSA Key modulus (value N) is compared against the x509 cert RSA Public Key modulus value N.  If both values do not match, x509 certificate and key are rejected with httpStatus 4XX.
5) Original certificate Chain validation functionality has been left in-place.
6) Ability to import x509v3 Certificates without SKI/AKI has been left in-place.

## Which TC components are affected by this PR?


## What is the best way to verify this PR? Please include manual steps or automated tests. 
### (If no tests are part of this PR, please provide explanation as to why no tests are included.)

Tests to perform for validation:
1) Import x509 certificate with mismatched RSA Private Key

Expected Result: x509 certificate and key are rejected and not saved in traffic vault.

2) Import x509 certificate with encrypted RSA Private Key

Expected Result: x509 certificate and key are rejected and not saved in traffic vault.

3) Attempt to import ECC x509 certificate and ECDSA private key 

Expected Result: x509 certificate and key are rejected and . not saved in traffic vault.

4) Import self-signed certificate for a delivery service with matching Private RSA key. 
    
5) Optional: Startup CiaB with this commit cherry-picked in.  

6) Run the unit tests in `deliveryservice/keys_test.go`
```
=== RUN   TestDecodePrivateKey_PKCS8_RSA_2048
--- PASS: TestDecodePrivateKey_PKCS8_RSA_2048 (0.00s)
=== RUN   TestDecodePrivateKey_PKCS1_RSA_2048
--- PASS: TestDecodePrivateKey_PKCS1_RSA_2048 (0.00s)
=== RUN   TestDecodePrivateKey_RSA_Encrypted_And_BadData
--- PASS: TestDecodePrivateKey_RSA_Encrypted_And_BadData (0.00s)
=== RUN   TestVerifyAndEncodeCertificate_BadData
--- PASS: TestVerifyAndEncodeCertificate_BadData (0.00s)
=== RUN   TestVerifyAndEncodeCertificate_SelfSigned_NO_SKIAKI_CertKeyPair
--- PASS: TestVerifyAndEncodeCertificate_SelfSigned_NO_SKIAKI_CertKeyPair (0.15s)
=== RUN   TestVerifyAndEncodeCertificate_SelfSignedCertKeyPair
--- PASS: TestVerifyAndEncodeCertificate_SelfSignedCertKeyPair (0.00s)
=== RUN   TestVerifyAndEncodeCertificate_SelfSignedCertKeyPair_MisMatched_PrivateKey
--- PASS: TestVerifyAndEncodeCertificate_SelfSignedCertKeyPair_MisMatched_PrivateKey (0.00s)
=== RUN   TestVerifyAndEncodeCertificate_CASignedCertKeyPair
--- PASS: TestVerifyAndEncodeCertificate_CASignedCertKeyPair (0.00s)
=== RUN   TestVerifyAndEncodeCertificate_CASigned_NO_SKIAKI_CertKeyPair
--- PASS: TestVerifyAndEncodeCertificate_CASigned_NO_SKIAKI_CertKeyPair (0.00s)
=== RUN   TestVerifyAndEncodeCertificate_CASigned_NO_SKIAKI_CertKeyPair_With_RootCA
--- PASS: TestVerifyAndEncodeCertificate_CASigned_NO_SKIAKI_CertKeyPair_With_RootCA (0.00s)
=== RUN   TestVerifyAndEncodeCertificate_CASignedCertKeyPair_With_RootCA
--- PASS: TestVerifyAndEncodeCertificate_CASignedCertKeyPair_With_RootCA (0.00s)
=== RUN   TestVerifyAndEncodeCertificate_ECC_CertificateKeyPair
--- PASS: TestVerifyAndEncodeCertificate_ECC_CertificateKeyPair (0.00s)
PASS
ok  	command-line-arguments	(cached)
```

## Check all that apply
